# 개요
- 참고 자료를 보셔도 좋지만, 코드 이해뿐만 아니라 데이터 전처리 처리에 대한 순서 / 법칙을 이해하면서 작성합시다!
- 코드에 대한 다양한 도전을 합시다 :) 
- 원본 데이터 : https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset/data

## 문제 1: 데이터 불러오기와 크기 확인
뇌졸중 예측 데이터셋 CSV 파일을 pandas를 사용하여 불러오고, 데이터프레임의 크기(행과 열의 개수)를 출력하는 코드를 작성하세요. (힌트: pd.read_csv 함수를 사용합니다.)

In [128]:
import pandas as pd # 판다스 데이터조작

df = pd.read_csv('superstore_orders.csv')
df

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
1     51676  Female  61.0             0              0          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5105  18234  Female  80.0             1              0          Yes   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
1     Self-employed          Rural             202.21   NaN     never smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
...             ...            ...                ...   ...              ...   
5105        Private          Urban              83.75   NaN     never smoked   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
5105       0  
5106       0  
5107       0  
5108       0  
5109       0  

[5110 rows x 12 columns]

## 문제 2: 데이터 미리보기 (앞부분 확인)
데이터가 잘 불러와졌는지 확인하기 위해 데이터프레임의 처음 5행을 출력하는 코드를 작성하세요. 이때 데이터의 각 열 이름도 함께 확인해보세요. 그리고 각 열의 의미를 적어보세요!  
그리고 특정 열이 어떤 값이 있는지 확인해보세요! (특정열.unique() 사용)

In [129]:
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

## 문제 3: 데이터 정보 확인 (데이터 타입과 결측값)
데이터프레임의 각 열에 대한 요약 정보를 출력하는 코드를 작성하세요. 이를 통해 각 열의 데이터 타입(dtype)과 누락된 값(null/결측값) 개수를 확인해보세요. (힌트: DataFrame.info() 메서드를 사용합니다.)

In [130]:
print(df.info(),"n/")



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB
None n/


## 문제 3-1: 데이터 정보 확인 (특정 컬럼 값 분포 확인)
hypertension, ever_married, avg_glucose_level, stroke 열의 값 분포를 확인하세요! (힌트: value_counts() 메서드를 사용합니다.)

In [131]:
df[['hypertension', 'ever_married', 'avg_glucose_level']].value_counts()

hypertension  ever_married  avg_glucose_level
0             Yes           91.68                5
                            72.49                4
              No            93.74                4
                            93.88                4
              Yes           114.32               4
                                                ..
                            57.47                1
                            57.52                1
                            57.57                1
                            57.59                1
1             Yes           271.74               1
Name: count, Length: 4546, dtype: int64

## 문제 4: 결측값 개수 확인
각 열별로 결측값(null)이 몇 개인지 합계로 확인하려 합니다. 이를 계산하는 코드를 작성하세요. (힌트: isnull() 또는 isna()와 sum() 함수를 활용합니다.)

In [132]:
df.isnull().sum()


id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

## 문제 5: id 열 삭제하기
id 열은 각 환자를 식별하기 위한 고유 식별자일 뿐, 예측 모델링에는 필요하지 않은 정보입니다. 데이터프레임에서 id 열을 제거하는 코드를 작성하세요.

In [133]:
df.drop(columns=["id"])

gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
1     Female  61.0             0              0          Yes  Self-employed   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
...      ...   ...           ...            ...          ...            ...   
5105  Female  80.0             1              0          Yes        Private   
5106  Female  81.0             0              0          Yes  Self-employed   
5107  Female  35.0             0              0          Yes  Self-employed   
5108    Male  51.0             0              0          Yes        Private   
5109  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0             Urban             228.69  36.6  formerly smoked       1  
1             Rural             202.21   NaN     never smoked       1  
2             Rural             105.92  32.5     never smoked       1  
3             Urban             171.23  34.4           smokes       1  
4             Rural             174.12  24.0     never smoked       1  
...             ...                ...   ...              ...     ...  
5105          Urban              83.75   NaN     never smoked       0  
5106          Urban             125.20  40.0     never smoked       0  
5107          Rural              82.99  30.6     never smoked       0  
5108          Rural             166.29  25.6  formerly smoked       0  
5109          Urban              85.28  26.2          Unknown       0  

[5110 rows x 11 columns]

## 문제 6: BMI 결측값 처리 (중앙값 대체)
bmi 열에 존재하는 201개의 결측값을 어떻게 처리할지 결정해야 합니다. 이번에는 중앙값(median)을 사용하여 이 결측치를 채우려고 합니다. bmi 열의 중앙값을 계산하고, 그 값으로 결측값들을 채우는 코드를 작성하세요.

In [134]:
isn_df = df['bmi'].isnull().sum()   #BMI행 결측값 갯수
fl_isn_df = df.dropna(subset='bmi') #BMI행 결측값 제거

bmi_md = df['bmi'].median()
df_md = df['bmi'].fillna(bmi_md)

print(df_md)




bmi_avg = df['bmi'].mean()                                              #실험용
bmi_sum_div = (df['bmi'].sum()) / (df['bmi'].count())                   #실험용
bmi_sum_div_total = (df['bmi'].sum()) / (df['bmi'].count()+isn_df)      #실험용
### 평균값 과 메디안 의 값은 다르다는것을 알게됨
## mean = 존재하는 데이터의 총합의 평균  
## median = 데이터 범주의 중앙값 (Min~Max Value의 중간 값)


0       36.6
1       28.1
2       32.5
3       34.4
4       24.0
        ... 
5105    28.1
5106    40.0
5107    30.6
5108    25.6
5109    26.2
Name: bmi, Length: 5110, dtype: float64


## 문제 7: 행 삭제(drop) 연습
bmi 값이 결측치(NaN) 이거나, bmi < 10 또는 bmi > 80인 행은 잘못된 데이터라고 가정합시다.  
이런 행들을 제거한 새로운 데이터프레임 df_clean을 만드세요. (원본 df는 그대로 둡니다.)  
제거 전/후 행 개수를 비교해서 출력해보세요.  

In [135]:
#제거 목표 : BMI < 10, BMI > 80 , BMI = NaN 

cond_index = df[ (df['bmi'] > 80) | (df['bmi'] < 10) ].index

df_drop = df.drop(cond_index).dropna()

print(df_drop,'n/')

print ("원본데이터 개수 : " , df['bmi'].shape)         #원본데이터 인포
print ("원본데이터 개수 : " , df_drop['bmi'].shape )    #드롭데이터 인포

         id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
5     56669    Male  81.0             0              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5104  14180  Female  13.0             0              0           No   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  former

## 문제 8: 조건 필터링 + 다중 정렬
나이 60세 이상이면서 (고혈압=1 또는 심장질환=1)인 행만 필터링해 생성  
avg_glucose_level 내림차순, 동률이면 bmi 오름차순으로 정렬  
상위 10명에 대해 id, age, avg_glucose_level, bmi, stroke만 출력  

In [136]:
cond_age = df[df['age']>60].index
df_age = df.loc[cond_age]

Taget = df_age.sort_values(by=['avg_glucose_level' , 'bmi'] , ascending=[False,True])

Taget

id  gender   age  hypertension  heart_disease ever_married  \
193   69112    Male  68.0             1              1          Yes   
3088  26267  Female  76.0             0              0          Yes   
4983  10995    Male  76.0             1              0          Yes   
3393  13155  Female  67.0             1              0          Yes   
135   71279  Female  71.0             0              0          Yes   
...     ...     ...   ...           ...            ...          ...   
3124  65413  Female  64.0             0              0          Yes   
2072  65252  Female  63.0             0              0          Yes   
659   50491    Male  78.0             0              0          Yes   
772   72547    Male  61.0             0              0          Yes   
4110  43028    Male  66.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
193         Private          Rural             271.74  31.1           smokes   
3088  Self-employed          Urban             267.61  27.9           smokes   
4983        Private          Rural             267.60  30.5     never smoked   
3393       Govt_job          Rural             263.56  26.3     never smoked   
135        Govt_job          Urban             263.32  38.7     never smoked   
...             ...            ...                ...   ...              ...   
3124        Private          Urban              55.64  43.4     never smoked   
2072       Govt_job          Rural              55.57  26.8  formerly smoked   
659   Self-employed          Urban              55.32  29.6           smokes   
772         Private          Rural              55.26  33.2          Unknown   
4110  Self-employed          Rural              55.23  28.9          Unknown   

      stroke  
193        1  
3088       0  
4983       0  
3393       0  
135        1  
...      ...  
3124       0  
2072       0  
659        0  
772        0  
4110       0  

[1304 rows x 12 columns]

## 문제 9: .loc vs .iloc 비교하기
데이터프레임 df에서 age, bmi, stroke 열만 뽑아서 확인해봅시다.
1. .loc를 사용해서 0~2번째 행의 age, bmi, stroke 값을 가져오세요.
2. .iloc를 사용해서 같은 행과 열을 가져오세요.
3. 두 결과가 같은지 눈으로 비교해보세요.

In [172]:
df_loc = df[['age','bmi','stroke']].loc[:2]
df_iloc = df[['age','bmi','stroke']].iloc[:3]

print(df_loc)
print('\n',df_iloc)
# .loc : 행의 '인덱스(index)' 이름이나 '불리언(boolean)' 조건을 사용하여 데이터를 선택 = 조건과 동일한 값 선택
# .iloc : 행과 열의 정수(숫자) 위치를 사용하여 데이터를 선택 = 데이터의 인덱스 위치값을 선택

#슬라이딩 END 값 
#    loc = END 값 포함
#   iloc = END 값 제외

    age   bmi  stroke
0  67.0  36.6       1
1  61.0   NaN       1
2  80.0  32.5       1

     age   bmi  stroke
0  67.0  36.6       1
1  61.0   NaN       1
2  80.0  32.5       1
